pkg/asset/installconfig/aws/validation.go

```diff
@@ -0,0 +1,153 @@
+		if _, ok := publicSubnets[id]; ok {
+			publicSubnetsIdx[id] = idx
+		}
+	}
+
+	allErrs = append(allErrs, validateSubnetCIDR(fldPath, privateSubnets, privateSubnetsIdx, networking.MachineCIDR)...)
+	allErrs = append(allErrs, validateSubnetCIDR(fldPath, publicSubnets, publicSubnetsIdx, networking.MachineCIDR)...)
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, privateSubnets, privateSubnetsIdx, "private")...)
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, publicSubnets, publicSubnetsIdx, "public")...)
+
+	privateZones := sets.NewString()
+	publicZones := sets.NewString()
+	for _, subnet := range privateSubnets {
+		privateZones.Insert(subnet.Zone)
+	}
+	for _, subnet := range publicSubnets {
+		publicZones.Insert(subnet.Zone)
+	}
+	if publish != types.InternalPublishingStrategy && !publicZones.IsSuperset(privateZones) {
+		errMsg := fmt.Sprintf("No public subnet provided for zones %s", privateZones.Difference(publicZones).List())
+		allErrs = append(allErrs, field.Invalid(fldPath, subnets, errMsg))
+	}
+
+	return allErrs
+}
+
+func validateMachinePool(ctx context.Context, meta *Metadata, fldPath *field.Path, platform *awstypes.Platform, pool *awstypes.MachinePool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(pool.Zones) > 0 && len(platform.Subnets) > 0 {
+		privateSubnets, err := meta.PrivateSubnets(ctx)
+		if err != nil {
+			return append(allErrs, field.Invalid(fldPath, meta.subnets, err.Error()))
+		}
+		privateZones := sets.NewString()
+		for _, subnet := range privateSubnets {
+			privateZones.Insert(subnet.Zone)
+		}
+		zones := sets.NewString(pool.Zones...)
+
+		if diff := zones.Difference(privateZones); diff.Len() > 0 {
+			errMsg := fmt.Sprintf("No subnets provided for zones %s", diff.List())
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("zones"), pool.Zones, errMsg))
+		}
+	}
+	return allErrs
+}
+
+func validateSubnetCIDR(fldPath *field.Path, subnets map[string]Subnet, idxMap map[string]int, machineCIDR *ipnet.IPNet) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for id, v := range subnets {
+		fp := fldPath.Index(idxMap[id])
+		cidr, _, err := net.ParseCIDR(v.CIDR)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(fp, id, err.Error()))
+			continue
+		}
+		if !machineCIDR.Contains(cidr) {
+			errMsg := fmt.Sprintf("CIDR range %s is outside of the MachineCIDR %s", v.CIDR, machineCIDR)
+			allErrs = append(allErrs, field.Invalid(fp, id, errMsg))
+			continue
+		}
+	}
+	return allErrs
+}
+
+func validateDuplicateSubnetZones(fldPath *field.Path, subnets map[string]Subnet, idxMap map[string]int, typ string) field.ErrorList {
+	var keys []string
+	for id := range subnets {
+		keys = append(keys, id)
+	}
+	sort.Strings(keys)
+
+	allErrs := field.ErrorList{}
+	zones := sets.String{}
+	for _, id := range keys {
+		subnet := subnets[id]
+		if zones.Has(subnet.Zone) {
+			errMsg := fmt.Sprintf("there is already a %s subnet for the zone %s", typ, subnet.Zone)
```

Comment:
I don't think we need to say "already", in case the user defined their subnets in a different order from our sorted comparison here (although I like us using a sorted comparison for stability).  How about `zones := make(map[string]string)`

```go
if conflictingSubnet, ok := zones[subnet.Zone]; ok {
  errMsg := fmt.Sprintf("%s subnet %s is also in zone %s", typ, conflictingSubnet, subnet.Zone)
```

Reply:
+1

Reply:
@wking fixed in https://github.com/openshift/installer/compare/a96182f3f4d7c7b69bd674bd918dd72a532f6ce4..9ce0bb39f5cbc2d581705794c8e16933cae350d4